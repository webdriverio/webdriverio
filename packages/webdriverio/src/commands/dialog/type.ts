/**
 * Returns dialog's type, can be one of `alert`, `beforeunload`, `confirm` or `prompt`.
 *
 * <example>
    :dialogType.js
    const type = await dialog.type();
 * </example>
 *
 * @alias dialog.type
 * @returns {string}  The type of the dialog
 */
// actual implementation is located in packages/webdriverio/src/dialog.ts
