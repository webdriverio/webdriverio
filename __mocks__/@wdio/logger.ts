import { vi } from 'vitest'

export const logMock = {
    error: vi.fn(),
    debug: vi.fn(),
    warn: vi.fn(),
    info: vi.fn(),
    trace: vi.fn(),
    progress: vi.fn()
}
const mock = () => logMock
mock.setLevel = vi.fn()
mock.setLogLevelsConfig = vi.fn()
mock.waitForBuffer = vi.fn()
mock.clearLogger = vi.fn()
export default mock
