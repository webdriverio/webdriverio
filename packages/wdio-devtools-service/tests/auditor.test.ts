import logger from '@wdio/logger'
import type { Trace } from '../src/gatherer/trace'

import Auditor from '../src/auditor'

let auditor

beforeEach(() => {
    auditor = new Auditor({} as unknown as Trace, [])
})

test('getMainThreadWorkBreakdown', async () => {
    expect(await auditor.getMainThreadWorkBreakdown()).toMatchSnapshot()
})

test('getDiagnostics', async () => {
    expect(await auditor.getDiagnostics()).toMatchSnapshot()
})

test('getDiagnostics failing', async () => {
    auditor._audit = jest.fn().mockReturnValue(Promise.resolve({}))
    expect(await auditor.getDiagnostics()).toBe(null)
})

test('getMetrics', async () => {
    expect(await auditor.getMetrics()).toMatchSnapshot()
})

test('getPerformanceScore', async () => {
    expect(await auditor.getPerformanceScore()).toMatchSnapshot()
})

test('getPerformanceScore: returns null if any of the metrics is not available', async () => {
    auditor._audit = jest.fn().mockReturnValueOnce(Promise.resolve({}))
    expect(await auditor.getPerformanceScore()).toBe(null)

    auditor._audit = jest.fn().mockReturnValueOnce(Promise.resolve({
        'first-contentful-paint': {
            score: 1
        },
    }))
    expect(await auditor.getPerformanceScore()).toBe(null)

    auditor._audit = jest.fn().mockReturnValueOnce(Promise.resolve({
        'first-contentful-paint': {
            score: 1
        },
        'speed-index': {
            score: 1
        },
    }))
    expect(await auditor.getPerformanceScore()).toBe(null)

    auditor.getMetrics = jest.fn().mockReturnValueOnce(Promise.resolve({
        'first-contentful-paint': {
            score: 1
        },
        'speed-index': {
            score: 1
        },
        'largest-contentful-paint': {
            score: 1
        },
    }))
    expect(await auditor.getPerformanceScore()).toBe(null)

    auditor._audit = jest.fn().mockReturnValueOnce(Promise.resolve({
        'first-contentful-paint': {
            score: 1
        },
        'speed-index': {
            score: 1
        },
        'largest-contentful-paint': {
            score: 1
        },
        'cumulative-layout-shift': {
            score: 1
        },
        'total-blocking-time': {
            score: 1
        }
    }))
    expect(await auditor.getPerformanceScore()).toBe(null)

    auditor._audit = jest.fn().mockReturnValue(Promise.resolve({
        'first-contentful-paint': {
            score: 1
        },
        'speed-index': {
            score: 1
        },
        'largest-contentful-paint': {
            score: 1
        },
        'cumulative-layout-shift': {
            score: 1
        },
        'total-blocking-time': {
            score: 1
        },
        'interactive': {
            score: 1
        },
    }))
    expect(await auditor.getPerformanceScore())
        .toEqual(expect.any(Number))
})

test('updateCommands', () => {
    const browser = { addCommand: jest.fn() }
    auditor.updateCommands(browser)

    expect(browser.addCommand)
        .toBeCalledWith('getMainThreadWorkBreakdown', expect.any(Function))
    expect(browser.addCommand)
        .toBeCalledWith('getDiagnostics', expect.any(Function))
    expect(browser.addCommand)
        .toBeCalledWith('getMetrics', expect.any(Function))
    expect(browser.addCommand)
        .toBeCalledWith('getPerformanceScore', expect.any(Function))
})

test('should not throw if no args passed', () => {
    const auditor = new Auditor()
    expect(auditor).toBeTruthy()
})

test('should throw if something fails', () => {
    const error = new Error('uups')
    const Audit = {
        defaultOptions: {},
        audit: jest.fn().mockImplementation(() => {
            throw error
        })
    }
    expect(auditor._audit(Audit)).toEqual({
        score: 0,
        error
    })
    expect(logger('').error).toBeCalledTimes(1)
})

test('should allow to audit PWA results', async () => {
    const auditor = new Auditor()
    expect(await auditor._auditPWA({ foo: 'bar' })).toMatchSnapshot()
    expect(await auditor._auditPWA(
        { foo: 'bar' },
        ['maskableIcon', 'serviceWorker']
    )).toMatchSnapshot()
})
