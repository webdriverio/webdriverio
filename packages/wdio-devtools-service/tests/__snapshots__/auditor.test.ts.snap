// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDiagnostics 1`] = `
Object {
  "mainDocumentTransferSize": 50074,
  "maxRtt": 145.65800000000002,
  "maxServerLatency": 4980.4710000000005,
  "numFonts": 5,
  "numRequests": 201,
  "numScripts": 36,
  "numStylesheets": 7,
  "numTasks": 110,
  "numTasksOver100ms": 0,
  "numTasksOver10ms": 1,
  "numTasksOver25ms": 0,
  "numTasksOver500ms": 0,
  "numTasksOver50ms": 0,
  "rtt": 19.016000000000002,
  "throughput": 569783.0945443822,
  "totalByteWeight": 3897521,
  "totalTaskTime": 56.882000000000026,
}
`;

exports[`getMainThreadWorkBreakdown 1`] = `
Array [
  Object {
    "duration": 25.485,
    "group": "styleLayout",
  },
  Object {
    "duration": 22.353,
    "group": "other",
  },
  Object {
    "duration": 4.353999999999999,
    "group": "scriptEvaluation",
  },
  Object {
    "duration": 1.678,
    "group": "parseHTML",
  },
  Object {
    "duration": 1.411,
    "group": "garbageCollection",
  },
  Object {
    "duration": 0.8830000000000001,
    "group": "scriptParseCompile",
  },
  Object {
    "duration": 0.718,
    "group": "paintCompositeRender",
  },
]
`;

exports[`getMetrics 1`] = `
Object {
  "cumulativeLayoutShift": 24347,
  "domContentLoaded": undefined,
  "estimatedInputLatency": 16,
  "firstCPUIdle": 23117,
  "firstContentfulPaint": 18090,
  "firstInteractive": 24117,
  "firstMeaningfulPaint": 23117,
  "firstPaint": undefined,
  "firstVisualChange": undefined,
  "largestContentfulPaint": 23117,
  "lastVisualChange": undefined,
  "load": undefined,
  "serverResponseTime": 105,
  "speedIndex": 20411,
  "timeToFirstByte": 105,
  "totalBlockingTime": 20411,
}
`;

exports[`getPerformanceScore 1`] = `0.8`;

exports[`should allow to audit PWA results 1`] = `
Object {
  "details": Object {
    "appleTouchIcon": Object {
      "details": Object {},
      "displayValue": "apple-touch-icon",
      "score": 1,
    },
    "contentWith": Object {
      "details": Object {},
      "displayValue": "content-width",
      "score": 1,
    },
    "isInstallable": Object {
      "details": Object {},
      "displayValue": "installable-manifest",
      "score": 1,
    },
    "maskableIcon": Object {
      "details": Object {},
      "displayValue": "maskable-icon",
      "score": 1,
    },
    "serviceWorker": Object {
      "details": Object {},
      "displayValue": "service-worker",
      "score": 1,
    },
    "splashScreen": Object {
      "details": Object {},
      "displayValue": "splash-screen",
      "score": 0,
    },
    "themedOmnibox": Object {
      "details": Object {},
      "displayValue": "themed-omnibox",
      "score": 1,
    },
    "viewport": Object {
      "details": Object {},
      "displayValue": "viewport",
      "score": 1,
    },
  },
  "passed": false,
}
`;

exports[`should allow to audit PWA results 2`] = `
Object {
  "details": Object {
    "maskableIcon": Object {
      "details": Object {},
      "displayValue": "maskable-icon",
      "score": 1,
    },
    "serviceWorker": Object {
      "details": Object {},
      "displayValue": "service-worker",
      "score": 1,
    },
  },
  "passed": true,
}
`;
